whelktool/scripts/cleanups/2022/06/lxl-3859-barn-ungdom.groovy

```diff
@@ -0,0 +1,51 @@
+/*
+I bibliografiska beskrivningar som har lokala formunderindelningen "barn- och ungdomslitteratur" i egenskapen Ämne ska
+den/dessa raderas och den bibliografiska beskrivningen ska kompletteras med genre/formtermen Barn- och ungdomslitteratur
+från systemet barngf i de fall en sådan term inte redan är inlagd.
+ */
+
+String where = """
+        collection = 'bib' AND deleted = false AND
+        data#>>'{@graph,1,instanceOf,subject}' LIKE '%barn- och ungdomslitteratur%'
+"""
+
+selectBySqlWhere(where) { data ->
+    boolean changed = false
+    def instance = data.graph[1]
+
+    Map subjectToPreserve = null
+
+    instance.instanceOf?.subject?.removeAll { subject ->
+
+        if (subject["@type"] == "ComplexSubject" && subject.inScheme?.containsKey("@id") && subject.inScheme["@id"] == "https://id.kb.se/term/barn") {
+            if (subject.termComponentList.removeAll { termComponent ->
+                return termComponent["@type"] == "GenreSubdivision" && termComponent.prefLabel == "barn- och ungdomslitteratur"
+                }) {
+                changed = true
+            }
+
+            // Only the main term still remaining? - Remove the whole subject
+            if (subject.termComponentList.size() == 1) {
+                subjectToPreserve = subject.termComponentList[0]
```

Comment:
Shouldn't all `subjectsToPreserve` (plural) be collected here (in a `Set<Map>`)? Otherwise, it seems that all but the last subject in e.g. https://libris.kb.se/7rkl60dk2q1r7tl#it would be dropped?

Reply:
Yepp! Good catch! :)